File: apps/clinical-engine/src/ehas2_clinical_engine/rule4/pediatric_overlay/evaluate_pediatric_overlay_adapter.py
from __future__ import annotations

from ..registry_paths import RULE4_CONTRACT_VERSION_PHASE9_PEDIATRIC_OVERLAY
from .overlay_code_validation import validate_pediatric_overlay_output_codes
from .pediatric_overlay_fingerprint_v1 import fingerprint_from_pediatric_overlay_output
from .resolve_slot_pediatric_overlay import resolve_slot_pediatric_overlay

RULE4_OVERLAY_GATE_OUTCOMES = ("PASS", "FAIL")

RULE4_PEDIATRIC_OVERLAY_STATUS_VALUES = (
    "NOT_EVALUATED",
    "NOT_APPLICABLE",
    "NOT_APPLICABLE_UNDER_HARD_STOP",
    "BLOCKED_BY_SAFETY",
    "BLOCKED_D13_HS",
    "AGE_UNRESOLVED",
    "PHASE8_AUTH_FAILED",
    "OVERLAY_APPLIED",
    "OVERLAY_BLOCKED",
    "OVERLAY_UNRESOLVED",
)


class Rule4PediatricOverlayAdapterValidationError(ValueError):
    code = "RULE4_PEDIATRIC_OVERLAY_ADAPTER_VALIDATION_FAILED"


def validate_pediatric_overlay_adapter_input(input_contract: dict) -> None:
    if input_contract.get("contract_version") != RULE4_CONTRACT_VERSION_PHASE9_PEDIATRIC_OVERLAY:
        raise Rule4PediatricOverlayAdapterValidationError(
            "contractVersion not supported for Phase 9 pediatric overlay"
        )
    if not input_contract.get("ruleset_version") or not input_contract.get("registry_version"):
        raise Rule4PediatricOverlayAdapterValidationError(
            "rulesetVersion and registryVersion required"
        )
    if input_contract.get("label") not in {"SYNTHETIC", "PRODUCTION"}:
        raise Rule4PediatricOverlayAdapterValidationError("label must be SYNTHETIC or PRODUCTION")


def evaluate_pediatric_overlay_adapter(input_contract: dict, context: dict | None = None) -> dict:
    context = context or {}
    validate_pediatric_overlay_adapter_input(input_contract)
    by_slot = {
        r["formula_slot_id"]: r for r in input_contract.get("slot_overlay_records") or []
    }
    meta = {
        "ruleset_version": input_contract["ruleset_version"],
        "registry_version": input_contract["registry_version"],
    }

    slot_resolutions = [
        resolve_slot_pediatric_overlay(
            by_slot.get(slot_id), slot_id, input_contract["label"], context, meta
        )
        for slot_id in input_contract.get("formula_slot_ids") or []
    ]

    reason_codes = sorted(
        set(
            [
                "PEDIATRIC_OVERLAY_SHADOW_DRAFT_ONLY",
                "PEDIATRIC_OVERLAY_NOT_PRESCRIPTION_ISSUANCE",
                *(c for s in slot_resolutions for c in s.get("reason_codes") or []),
                *(
                    ["PRODUCTION_PEDIATRIC_OVERLAY_NOT_EVALUATED"]
                    if input_contract.get("label") == "PRODUCTION"
                    else []
                ),
            ]
        )
    )

    limitation_codes = sorted(
        set(
            [
                "SHADOW_PEDIATRIC_DRAFT_OVERLAY_ONLY",
                *(c for s in slot_resolutions for c in s.get("limitation_codes") or []),
            ]
        )
    )

    safety_gate = context.get("safety_gate") or {}
    safety_flags = {
        "d13_hs_active": bool(safety_gate.get("d13_hard_stop_active")),
        "patient_wide_hold": bool(safety_gate.get("patient_wide_hold")),
        "urgent_escalation_required": bool(safety_gate.get("urgent_escalation_required")),
    }

    core = {
        "contract_version": input_contract["contract_version"],
        "ruleset_version": input_contract["ruleset_version"],
        "registry_version": input_contract["registry_version"],
        "execution_status": "NOT_IMPLEMENTED",
        "automatic_pediatric_overlay_runtime": False,
        "automatic_prescription_issuance_runtime": False,
        "prescription_issue_allowed": False,
        "final_doctor_approval_required": True,
        "current_runtime_potency_delta": "NONE",
        "slot_resolutions": slot_resolutions,
        "reason_codes": reason_codes,
        "limitation_codes": limitation_codes,
        "deterministic_pediatric_overlay_fingerprint": "",
    }

    validate_pediatric_overlay_output_codes(core)
    core["deterministic_pediatric_overlay_fingerprint"] = fingerprint_from_pediatric_overlay_output(
        core, safety_flags
    )
    return core

File: packages/clinical-contracts/src/rule4/pediatricOverlay/evaluatePediatricOverlayAdapter.ts
import { RULE4_CONTRACT_VERSION_PHASE9_PEDIATRIC_OVERLAY } from '../version.js';
import { fingerprintFromPediatricOverlayOutput } from './pediatricOverlayFingerprintV1.js';
import { resolveSlotPediatricOverlay } from './resolveSlotPediatricOverlay.js';
import { validatePediatricOverlayOutputCodes } from './overlayCodeValidation.js';
import type {
  Rule4PediatricOverlayAdapterInput,
  Rule4PediatricOverlayAdapterOutput,
  Rule4PediatricOverlayEvaluationContext,
} from './types.js';

export class Rule4PediatricOverlayAdapterValidationError extends Error {
  readonly code = 'RULE4_PEDIATRIC_OVERLAY_ADAPTER_VALIDATION_FAILED';

  constructor(message: string) {
    super(message);
    this.name = 'Rule4PediatricOverlayAdapterValidationError';
  }
}

export function validatePediatricOverlayAdapterInput(
  input: Rule4PediatricOverlayAdapterInput,
): void {
  if (input.contractVersion !== RULE4_CONTRACT_VERSION_PHASE9_PEDIATRIC_OVERLAY) {
    throw new Rule4PediatricOverlayAdapterValidationError(
      'contractVersion not supported for Phase 9 pediatric overlay',
    );
  }
  if (!input.rulesetVersion || !input.registryVersion) {
    throw new Rule4PediatricOverlayAdapterValidationError(
      'rulesetVersion and registryVersion required',
    );
  }
  if (input.label !== 'SYNTHETIC' && input.label !== 'PRODUCTION') {
    throw new Rule4PediatricOverlayAdapterValidationError('label must be SYNTHETIC or PRODUCTION');
  }
}

export function evaluatePediatricOverlayAdapter(
  input: Rule4PediatricOverlayAdapterInput,
  context: Rule4PediatricOverlayEvaluationContext = {},
): Rule4PediatricOverlayAdapterOutput {
  validatePediatricOverlayAdapterInput(input);
  const bySlot = new Map(input.slotOverlayRecords.map((r) => [r.formulaSlotId, r]));
  const meta = { rulesetVersion: input.rulesetVersion, registryVersion: input.registryVersion };

  const slotResolutions = input.formulaSlotIds.map((id) =>
    resolveSlotPediatricOverlay(bySlot.get(id), id, input.label, context, meta),
  );

  const reasonCodes = [
    'PEDIATRIC_OVERLAY_SHADOW_DRAFT_ONLY',
    'PEDIATRIC_OVERLAY_NOT_PRESCRIPTION_ISSUANCE',
    ...slotResolutions.flatMap((s) => s.reasonCodes),
    ...(input.label === 'PRODUCTION' ? ['PRODUCTION_PEDIATRIC_OVERLAY_NOT_EVALUATED'] : []),
  ];

  const limitationCodes = [
    'SHADOW_PEDIATRIC_DRAFT_OVERLAY_ONLY',
    ...slotResolutions.flatMap((s) => s.limitationCodes),
  ];

  const safetyFlags = {
    d13HsActive: Boolean(context.safetyGate?.d13HardStopActive),
    patientWideHold: Boolean(context.safetyGate?.patientWideHold),
    urgentEscalationRequired: Boolean(context.safetyGate?.urgentEscalationRequired),
  };

  const core: Rule4PediatricOverlayAdapterOutput = {
    contractVersion: input.contractVersion,
    rulesetVersion: input.rulesetVersion,
    registryVersion: input.registryVersion,
    executionStatus: 'NOT_IMPLEMENTED',
    automaticPediatricOverlayRuntime: false,
    automaticPrescriptionIssuanceRuntime: false,
    prescriptionIssueAllowed: false,
    finalDoctorApprovalRequired: true,
    currentRuntimePotencyDelta: 'NONE',
    slotResolutions,
    reasonCodes: [...new Set(reasonCodes)].sort(),
    limitationCodes: [...new Set(limitationCodes)].sort(),
    deterministicPediatricOverlayFingerprint: '',
  };

  validatePediatricOverlayOutputCodes(core);

  const fp = fingerprintFromPediatricOverlayOutput(core, safetyFlags);
  return { ...core, deterministicPediatricOverlayFingerprint: fp };
}
